crates/ruff_workspace/src/cli.rs

```diff
@@ -0,0 +1,157 @@
+#[cfg(feature = "clap")]
+pub mod clap_completion {
+    use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
+    use itertools::Itertools;
+    use std::str::FromStr;
+
+    use crate::{
+        options::Options,
+        options_base::{OptionField, OptionSet, OptionsMetadata, Visit},
+    };
+
+    #[derive(Default)]
+    struct CollectOptionsVisitor {
+        values: Vec<(String, String)>,
+        parents: Vec<String>,
```

Comment:
Not too important but we can avoid some allocations by using `&'static str`
```suggestion
        values: Vec<(String, &'static str)>,
        parents: Vec<&'static str>,
```

This does require changing the signature of the `Visit` methods to `&'static str` but I think that should be easily possible.

Reply:
There are 4 structs implementing this Trait, so I didn't make the change to limit the scope of this PR. If you'd like me to go ahead and change all of them I can. But, tbh, this code is ran once per ruff install/update, I am pretty sure the difference is negligible